TombEngine/Objects/TR2/Entity/tr2_sword_guardian.cpp

```diff
@@ -12,210 +13,372 @@
 #include "Game/misc.h"
 #include "Sound/sound.h"
 #include "Specific/level.h"
-
-using namespace TEN::Math::Random;
+#include "Specific/setup.h"
 
 namespace TEN::Entities::Creatures::TR2
 {
+	constexpr auto SWORD_GUARDIAN_FLY_TURN_RATE_MAX = ANGLE(7.0f);
+	constexpr auto SWORD_GUARDIAN_WALK_TURN_RATE_MAX = ANGLE(9.0f);
+
 	const auto SwordBite = BiteInfo(Vector3(0.0f, 37.0f, 550.0f), 15);
 
-	void SwordGuardianFly(ItemInfo* item)
+	constexpr auto SWORD_GUARDIAN_ATTACK_RANGE = SQUARE(SECTOR(1));
+	constexpr auto SWORD_GUARDIAN_WALK_ATTACK_RANGE = SQUARE(SECTOR(2));
+
+	constexpr auto SWORD_GUARDIAN_SWAPMESH_TIME = 3;
+	constexpr auto SWORD_GUARDIAN_ATTACK_DAMAGE = 300;
+
+	enum SwordGuardianState
+	{
+		SWORD_GUARDIAN_STATE_STOP = 1,
+		SWORD_GUARDIAN_STATE_WALK = 2,
+		SWORD_GUARDIAN_STATE_ATTACK_FRONT_AIM = 3,
+		SWORD_GUARDIAN_STATE_ATTACK_FRONT = 4,
+		SWORD_GUARDIAN_STATE_ATTACK_HORIZONTAL_AIM = 5,
+		SWORD_GUARDIAN_STATE_ATTACK_HORIZONTAL = 6,
+		SWORD_GUARDIAN_STATE_WAIT = 7,
+		SWORD_GUARDIAN_STATE_FLY = 8,
+		SWORD_GUARDIAN_STATE_AWAKE = 9,
+		SWORD_GUARDIAN_STATE_WALK_ATTACK_AIM = 10,
+		SWORD_GUARDIAN_STATE_WALK_ATTACK = 11,
+		SWORD_GUARDIAN_STATE_DEATH = 12
+	};
+
+	// TODO: Found all name, also found some which is more appropriate.
+	enum SwordGuardianAnim
+	{
+		SWORD_GUARDIAN_ANIM_AWAKE = 0
+	};
+
+	static void SpawnSpearGuardianSmoke(const Vector3i& pos, int roomNumber)
+	{
+		auto& smoke = *GetFreeParticle();
+
+		bool isUnderwater = TestEnvironment(ENV_FLAG_WATER, pos, roomNumber);
+		auto sphere = BoundingSphere(pos.ToVector3(), 16);
+		auto effectPos = Random::GeneratePointInSphere(sphere);
+
+		smoke.on = true;
+		smoke.blendMode = BLEND_MODES::BLENDMODE_ADDITIVE;
+
+		smoke.x = effectPos.x;
+		smoke.y = effectPos.y;
+		smoke.z = effectPos.z;
+		smoke.xVel = Random::GenerateInt(-BLOCK(0.5f), BLOCK(0.5f));
+		smoke.yVel = Random::GenerateInt(-BLOCK(1 / 8.0f), BLOCK(1 / 8.0f));
+		smoke.zVel = Random::GenerateInt(-BLOCK(0.5f), BLOCK(0.5f));
+
+		if (isUnderwater)
```

Comment:
Is sword guardian ever underwater?

Reply:
No, it was copy-pasted, and just in case, i let it there.